VSR: In-place upgrades

sentientwaffle opened the pull request:
To ensure storage determinism over state machine + LSM modifications, code upgrades are coordinated (via consensus) across the cluster. Multiple versions of the code will be compiled into a single `tigerbeetle` binary so that the replica itself can decide which version should be running, and deploy its own upgrades. This approach optimizes for safely (eliminating a large category of storage nondeterminism) and operational simplicity.

1. Replicas share which releases they are willing to upgrade into over `ping` messages.
2. When a `status=normal` primary detects a majority cluster that is willing to upgrade to a particular release, it triggers an upgrade...
3. Primary sends a `command=request operation=upgrade client=0` to itself, bearing the target release `UpgradeRequest { release: u16 }`.
4. When `operation=upgrade` commits, the primary sends itself another, until both:
  - `commit_min == op_checkpoint_next_trigger`, AND
  - All messages in the last bar are `operation=upgrade`.
5. At this point, the replica checkpoints, increasing its `CheckpointState.release` to the new value.
6. When the checkpoint completes, the replicas sees that it is running an older release (`self.release < self.superblock.working.vsr_state.checkpoint.release`) and attempts to execute the new release.

Other edge cases to keep in mind:
- If the above steps are interrupted by a view change: The first `commit_upgrade` sets `self.upgrade_release`, so that the new primary resumes sending the padding `operation=upgrade`s. (This is triggered by `on_upgrade_timeout()`.)
- If we send a padding upgrade request to ourself, then immediately drop it due to clock-not-synchronized, `on_upgrade_timeout` will keep trying.
- When a replica finishes state sync, it checks whether it needs to transition to a new release.

`ignore_request()` does not allow any request to queue behind an `operation=upgrade`:
- This simplifies the mental model + invariants of upgrades.
- This ensures that we will reach the "last bar is all checkpoints" soon after an upgrade is triggered.

Note that all `operation=upgrade` messages set `client=0`, which means they cannot displace any real clients' sessions from the sessions table.

Clients also have a release number, which they include in their message headers.
The replica has a minimum client release which they accept requests from. Messages from clients running a release which is less-than that minimum, or greater than the cluster's current release, will be rejected by a `command=eviction` message.

sentientwaffle committed:
Test/ReplySequence: Set up ReplySequence to allow skipped prepares

The `Simulator`'s `ReplySequence` assembles replies (from the cluster to all clients) into commit order.
(Replies may arrive out of order, but the workload checker needs to process them in commit order so that the workload can correctly reproduce the cluster's sequence of events.)

Right now `ReplySequence` has a comprehensive view of replies: currently all `command=request`s originate from a client (`client_id ≠ 0`).

But shortly we will introduce request types that originate from the primary itself (`client = 0`):

- upgrade operations
- expiry messages, to drive transfer timeouts.

As such, this commit updates `ReplySequence` so that we will be able to skip past upgrade/expiry operations that the clients will never receive a reply for. This will work by changing the `else` branch of the `on_cluster_reply` loop to:

    if (prepare_header.client == 0) {
        simulator.reply_op_next += 1;
    } else {
        break;
    }

sentientwaffle committed:
VSR: Support internal client=0 requests/prepares

Allow the primary of the cluster to send requests to itself.
These have some differences from normal client requests:

- `header.client == 0`
- Only specific operations are eligible. (None in this commit).
- No session
- No reply

Note that this doesn't increase the pipeline size, since client=0 messages should not be common.

sentientwaffle committed:
Test/StateChecker: Verify commits' release versions

sentientwaffle committed:
Test: Replica.Options.test_context

Move `test_context` into `Replica.open()`'s options.
Replicas will need to transition ("execute") to a different release from within `Replica.open()` (synchronously). For tests, this callback is implemented by `testing/cluster.zig`, and needs the `test_context` so that it can retrieve the `cluster` reference.

sentientwaffle committed:
Test: Cluster.Options.releases

sentientwaffle committed:
Test/Cluster: Implement Replica.release_execute

sentientwaffle committed:
VSR: Downgrade during Replica.open()

When a replica first starts (`./tigerbeetle start`), the multiversion binary will execute the highest available version.

If we open the superblock and see that our code release is higher than the superblock `CheckpointState`'s release, then we will `release_execute()` into the superblock's release.
(If that release is unavailable, that is an operator error – we will panic.)

sentientwaffle committed:
VSR/SuperBlock: Set release during .checkpoint()

sentientwaffle committed:
VSR: Add consensus-based in-place upgrades

To ensure storage determinism over state machine + LSM modifications, code upgrades are coordinated (via consensus) across the cluster. Multiple versions of the code will be compiled into a single `tigerbeetle` binary so that the replica itself can decide which version should be running, and deploy its own upgrades. This approach optimizes for safely (eliminating a large category of storage nondeterminism) and operational simplicity.

1. Replicas share which releases they are willing to upgrade into over `ping` messages.
2. When a `status=normal` primary detects a majority cluster that is willing to upgrade to a particular release, it triggers an upgrade...
3. Primary sends a `command=request operation=upgrade client=0` to itself, bearing the target release `UpgradeRequest { release: u16 }`.
4. When `operation=upgrade` commits, the primary sends itself another, until both:
  - `commit_min == op_checkpoint_next_trigger`, AND
  - All messages in the last bar are `operation=upgrade`.
5. At this point, the replica checkpoints, increasing its `CheckpointState.release` to the new value.
6. When the checkpoint completes, the replicas sees that it is running an older release (`self.release < self.superblock.working.vsr_state.checkpoint.release`) and attempts to execute the new release.

Other edge cases to keep in mind:
- If the above steps are interrupted by a view change: The first `commit_upgrade` sets `self.upgrade_release`, so that the new primary resumes sending the padding `operation=upgrade`s. (This is triggered by `on_upgrade_timeout()`.)
- If we send a padding upgrade request to ourself, then immediately drop it due to clock-not-synchronized, `on_upgrade_timeout` will keep trying.
- When a replica finishes state sync, it checks whether it needs to transition to a new release.

`ignore_request()` does not allow any request to queue behind an `operation=upgrade`:
- This simplifies the mental model + invariants of upgrades.
- This ensures that we will reach the "last bar is all checkpoints" soon after an upgrade is triggered.

Note that all `operation=upgrade` messages set `client=0`, which means they cannot displace any real clients' sessions from the sessions table.

Clients also have a release number, which they include in their message headers.
The replica has a minimum client release which they accept requests from. Messages from clients running a release which is less-than that minimum, or greater than the cluster's current release, will be rejected by a `command=eviction` message.

sentientwaffle committed:
Test/Replica: Test basic upgrade cases

sentientwaffle committed:
VSR: Test for upgrade after state sync

sentientwaffle committed:
VOPR: Inject upgrades in simulator

The only "surprising" thing here is that `Simulator.done()` should wait until any in-progress updates are done.

Currently we don't ever "withdraw" old releases when they are no longer used.
That is less useful from a test surface perspective, but we can add it later.

sentientwaffle committed:
VOPR: Clean up loop

sentientwaffle committed:
VSR: Send command=commit before checkpoint begins

Send a commit message immediately before a primary begins to checkpoint, rather than waiting for our commit timer.

This is useful when this checkpoint is an upgrade, since we will need to restart into the new version. We want all the replicas to restart in parallel (as much possible) rather than in sequence (to minimize downtime due to recovery/replay).

sentientwaffle commented on src/simulator.zig:
```diff
@@ -666,36 +727,55 @@ pub const Simulator = struct {
-                    commit.request.body(),
-                    commit.reply.body(),
-                );
+        while (!simulator.reply_sequence.empty()) {
```
The `Simulator`'s `ReplySequence` assembles replies (from the cluster to all clients) into commit order.
(Replies may arrive out of order, but the workload checker needs to process them in commit order so that the workload can correctly reproduce the cluster's sequence of events.)

Right now `ReplySequence` has a comprehensive view of replies: currently all `command=request`s originate from a client (`client_id ≠ 0`).

But upgrades originate from the primary itself (`client = 0`)! (Expiry messages for driving transfer timeouts will too.)

As such, this commit updates `ReplySequence` so that we will be able to skip past upgrade/expiry operations that the clients will never receive a reply for.

sentientwaffle commented on src/simulator.zig:
```diff
@@ -167,6 +175,8 @@ pub fn main() !void {
         .replica_crash_stability = random.uintLessThan(u32, 1_000),
         .replica_restart_probability = 0.0002,
         .replica_restart_stability = random.uintLessThan(u32, 1_000),
+        .replica_release_advance_probability = 0.0001,
+        .replica_release_catchup_probability = 0.001,
```
These are separate because we want "probability that the cluster decides to upgrade" to be lower than "probability that a replica can catch up to an already-upgraded cluster".

sentientwaffle commented on src/vsr.zig:
```diff
@@ -220,6 +220,8 @@ pub const Operation = enum(u8) {
     /// The value 3 is reserved for reconfiguration request.
     reconfigure = 3,
+    /// The value 4 is is reserved for release-upgrade requests.
+    upgrade = 4,
```
I was tempted to name this "release" to keep up the "re*" trend, but "upgrade" is more precise and less overloaded. :stuck_out_tongue:

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -869,6 +896,8 @@ pub fn ReplicaType(
             release_client_min: u16,
             releases_bundled: []const u16,
+            release_execute: *const fn (replica: *Self, release: u16) void,
+            test_context: ?*anyopaque,
```
Replicas will need to transition ("execute") to a different release from within `Replica.open()` (synchronously). For tests, this callback is implemented by `testing/cluster.zig`, and needs the `test_context` so that it can retrieve the `cluster` reference.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2969,6 +2997,72 @@ pub fn ReplicaType(
+                    // to ensure that the upgraded cluster can both commit and view-change.
+                    if (release_replicas >= vsr.quorums(self.replica_count).majority) {
+                        release_target = release;
+                    }
```
Replicas advertise (in `ping` messages) the list of release versions their binary supports. When the primary sees that enough replicas have a newer version available, it kicks off the upgrade process.

How many is "enough"? The lower bound is `max(quorum_commit, quorum_view_change) == quorum_majority`, since we want the quorum in the new version to be able to both commit and change views.

I am tempted to set the threshold to `self.replica_count`. In practice, not upgrading all replicas together would probably be an operator error which leads to lagging replicas that then need to state sync.

...But I'm worried about the case of: A replica is broken/recovering, and the operator may _need_ to upgrade anyway. (e.g. to roll out a hotfix quickly)

matklad replied:
Maybe `self.replica_count - 1`? In any case, I suggest writing 


```
vsr.quorums(self.replica_count).upgrade
```

even if that ends up being equal to majority.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -3970,6 +4101,49 @@ pub fn ReplicaType(
+                assert(self.upgrade_release == null);
+            } else {
+                if (self.upgrade_release) |upgrade_release| {
+                    assert(upgrade_release == request.release);
```
Once an "upgrade to release X" is committed, the replica cannot change its mind and decide to commit "upgrade to release Y" (for any Y≠X) until release X has checkpointed and the replica is running release X.

(We could ease this restriction later if necessary, but I think it is a helpful simplification for my mental model if nothing else.)

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4482,14 +4711,14 @@ pub fn ReplicaType(
             if (self.client_sessions.get(message.header.client)) |entry| {
                 assert(entry.header.command == .reply);
                 assert(entry.header.client == message.header.client);
+                assert(entry.header.client != 0);
```
`client=0` doesn't have a session.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4687,12 +4925,12 @@ pub fn ReplicaType(
             if (self.pipeline.queue.message_by_client(message.header.client)) |pipeline_message| {
                 assert(pipeline_message.header.command == .request or
                     pipeline_message.header.command == .prepare);
+                assert(message.header.client != 0);
```
Since `client=0` has no session, the `pipeline.queue.message_by_client` case is already covered for client 0 in `ignore_request_message_duplicate()`.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -8761,6 +9092,88 @@ pub fn ReplicaType(
+            maybe(self.pipeline.queue.message_by_operation(.upgrade) != null);
+        }
+
+        fn send_request_to_self(self: *Self, operation: vsr.Operation, body: []const u8) void {
```
I wrote this thinking that the [pulse messages](https://github.com/tigerbeetle/tigerbeetle/pull/1657) might be able to use this same function, but it looks like pulses are injected in a very different way.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -5260,8 +5498,6 @@ pub fn ReplicaType(
             assert(!self.pipeline.queue.prepare_queue.full());
             self.pipeline.queue.verify();
 
-            assert(!self.ignore_request_message(request.message));
```
This assertion doesn't play nicely with upgrades. Consider:

1. Queue request X, operation=whatever
2. Queue request X+1, operation=upgrade
3. A commit finishes, pops request X from the queue and tries to prepare it...

But it hits `on_request: ignoring (upgrade queued)` here, even though the request precedes the upgrade so it should work.

sentientwaffle committed:
VSR: Remove overstrict assertions

We might have received the view_headers via a `start_view` but not yet written them to the superblock.

sentientwaffle committed:
VSR: upgrade_quorum=max(majority,replica_count-1)

matklad commented on src/vsr/replica.zig:
```diff
@@ -9434,6 +9843,20 @@ const PipelineQueue = struct {
         return message;
     }
 
+    fn message_by_operation(pipeline: PipelineQueue, operation: vsr.Operation) ?*const Message {
```
This feels like an error-prone API, because it is not specified _which_ message is returned in case several match. 

Maybe `fn operation_pipelined(pipeline: PipelineQueue, operation: vsr.Operation) bool`?

Internally, it'll still compute the actual message and assert that we do have a message with the specifed op, but the call site won't be able to rely on impl detail of which message is returned.

matklad commented on src/vsr/replica.zig:
```diff
@@ -8313,6 +8566,80 @@ pub fn ReplicaType(
+                    // If v1 and v2 produce different results when executing op=X, then an assertion
+                    // will trip (as v2's reply doesn't match v1's in the client sessions).
+                    assert(found_upgrade < constants.lsm_batch_multiple);
+                    maybe(found_upgrade > 0);
```
To explain this maybe: if a cluster is idle, the op is between a checkpoint and its trigger, and the cluster receives an upgrade request, then:

* we want to append the request to log immediately
* but we want to pulse a checkpoint worth of upgrades up to the _next_ checkpoint, because the first critical bar of the current checkpoint is already poisoned by a non-upgrade request.

matklad replied:
Wait, no something doesn't add up here? If there is an upgrade committed, anything on top of it should also be an upgrade? 

That is, across the _entire_ loop, `found_upgrade` could be greater than 0 but less than `lsm_batch_multiple`. However, if there's a non-upgrade op in this batch, it has to be the first op, no? 

Am I misunderstanding the big picture here?

sentientwaffle replied:
Ah, you're right, this should be
```
assert(found_upgrade == 0);
```

This is a relic from an earlier iteration which allowed non-upgrades to queue up behind the upgrades if no upgrade had actually committed yet.

matklad commented on src/vsr.zig:
```diff
@@ -543,6 +545,15 @@ test "ReconfigurationRequest" {
+        assert(@sizeOf(UpgradeRequest) == 2);
+        assert(stdx.no_padding(UpgradeRequest));
+    }
+};
```
Hm, won't this hit the same alignment issue as pings?

matklad commented on src/vsr/replica.zig:
```diff
@@ -8173,6 +8419,13 @@ pub fn ReplicaType(
             ));
 
             self.commit_min = self.superblock.working.vsr_state.checkpoint.commit_min;
+            self.upgrade_release = null;
```
Maybe _not_ null this out? My understanding is that, when we hit this case, the following code is essentially `noreturn`. So, if we _don't_ null out here, we'll trip assertions if `release_transition` is buggy and doesn't actually reset the entire state!

matklad commented on src/vsr/replica.zig:
```diff
@@ -8761,6 +9088,88 @@ pub fn ReplicaType(
+            }));
+        }
+
+        fn send_request_upgrade_to_self(self: *Self) void {
```
Do we guard anywhere that nobody else can send us an upgrade request? I think we can check that in `on_message_from_bus`?

sentientwaffle replied:
We do check that `operation == upgrade` implies `client == 0`, but verifying in `on_message_from_bus` is a good idea!

matklad replied:
Yeah, I was thinking about an evil client sending a message with `client = 0` "manually"

sentientwaffle committed:
VSR: Pad operation=upgrade body to u256 multiple

See https://github.com/tigerbeetle/tigerbeetle/issues/1518

sentientwaffle committed:
VSR: Improve operation=upgrade assertions

sentientwaffle committed:
VSR: Replace message_by_operation with contains_operation

Safer API, since there is no worry about which message is returned if there are multiple matches.

sentientwaffle committed:
VSR: Don't reset Replica.upgrade_release after state sync

If we just synced to a newer checkpoint, and `self.upgrade_release ≠ null`, then by definition our superblock's release has increased, so we will transition to the new release. No need to reset `upgrade_release`, we can assert instead.

sentientwaffle commented:
(Rebased atop main to resolve a few git conflicts.)

matklad commented on src/tigerbeetle/main.zig:
```diff
@@ -325,6 +326,21 @@ const Command = struct {
     }
 };
 
+fn replica_release_execute(replica: *Replica, release: u16) void {
```
```suggestion
fn replica_release_execute(replica: *Replica, release: u16) noreturn {
```

matklad commented on src/vsr/replica.zig:
```diff
@@ -1339,7 +1380,9 @@ pub fn ReplicaType(
                     const releases_all = std.mem.bytesAsSlice(u16, message.body());
                     const releases = releases_all[0..message.header.release_count];
```
should assert that releases _past_ `release_count` are all zeroes

matklad commented on src/vsr/replica.zig:
```diff
@@ -3974,6 +4106,50 @@ pub fn ReplicaType(
+                if (self.upgrade_release) |upgrade_release| {
+                    assert(upgrade_release == request.release);
+                } else {
+                    self.upgrade_release = request.release;
```
Assert that anything that's pipelined at this point is an upgrade. 

The subtlty here is that, when a view change happens during upgrade, the new primary can't prepare anything until it commits backlog. 

And maybe assert in `PipelineQueue.verify` that nothing can be pielined past an upgrade?

sentientwaffle replied:
Ah.. earlier in the function I had the assert:

```
            if (self.pipeline == .queue) {
                assert(self.pipeline.queue.prepare_queue.count == 1);
                assert(self.pipeline.queue.request_queue.empty());
            }
```

But this is not actually correct... If we are a new primary (still in status=view_change) committing the ops in our pipeline, we may have multiple upgrades!

matklad commented on src/vsr/replica_test.zig:
```diff
@@ -1240,6 +1247,108 @@ test "Cluster: prepare beyond checkpoint trigger" {
+        t.replica(.R_).pass(.__, .bidirectional, .prepare_ok);
+        t.run();
+        t.run();
+        t.run();
```
Do we need all three for the release to propagate?

sentientwaffle replied:
Ah, oops! It used to need those b/c I originally set the `upgrade_timeout` much higher.

matklad commented on src/simulator.zig:
```diff
@@ -797,12 +890,21 @@ pub const Simulator = struct {
-                        continue;
-                    }
+                        simulator.options.replica_release_catchup_probability,
+                    ));
```
Not the prettiest expression! 

Can't suggest any local improvements, perhaps its time to introuce `tick_crash_up` and `tick_crash_down` functions to save on indentation here

sentientwaffle replied:
Good idea. I also modified `tick_crash_down` so that we recompute `recoverable_count` before restarting a replica, rather than tracking it incrementally.

matklad reviewed (commented):
More or less finished review! Looks good to me module a couple of missing asserts!

sentientwaffle committed:
VSR: Upgrades' code review

sentientwaffle committed:
VOPR: Split tick_crash into tick_crash_{up,down}

Also recompute `recoverable_count` every time we restart a replica, rather than tracking it incrementally. In addition to being simpler, this seems less error-prone.

matklad commented:
zig fmt fails!

sentientwaffle committed:
VSR: Allow AOF recover with client=0 messages
